### DISCUSSION TITLE
Connect to background process running on different port

### DISCUSSION THREAD

**Discussion by bcpeinhardt** [emilk/egui#2353] [Q&A]
Hello! I could use some help, as I'm completely new to non web based GUI programming.

I am working on a UI testing DSL that executes against a running selenium/webdriver process. I'm trying to build a GUI for it using egui, and I'm having trouble doing a few things :

1. Starting, managing, and closing a background process on a separate port from egui.
2. Executing chunks of async code.

What is the most straightforward way in egui to create a UI element that will trigger the execution of some async code and get notified when it completes?
Also, is there a standard way to manage background processes with eframe/egui?

For more on what I'm trying to do see: https://github.com/bcpeinhardt/schnauzerUI/blob/main/src/bin/gui.rs

**Comment by coderedart** [comment]
I am not aware of any standard way to manage background processes.

But for async code, you might want to look at the [README](https://github.com/emilk/egui/blob/master/README.md#how-do-i-use-egui-with-async).

If most of these async tasks are atomic / one shot (just start in background and give me result when done), just pushing them all  into the background using something like `poll_promise` and inserting the promise into a hashmap. you can just iterate over the hashmap every frame and check which one of them promises are finished. 

If you are dealing with a task that runs for a while and wants to communicate with the foreground thread (UI) like sending percentage of completion etc.. then prefer using channels. again, just push the task into background with a send part of channel and use the receiver part to get the latest updates from the background task. if you need to communicate back, keep two channels, the first is for background task to send updates to foreground and the second is for foreground UI thread to send updates/commands to background thread. every frame, just iterate over all the receiver channels, check all the updates received, and send anything back to background tasks using senders.
